### PR TITLE
[Polaris Lighthouse] Add migration to insert disable comments for @shopify/stylelint-polaris

### LINKED ISSUE #7504: Create a rollout strategy
see below for proposed strategy

### PR DESCRIPTION
**Pull request by sophschneider**
**NOTE: even though this PR is independent of the [stylelint-polaris v5](https://github.com/Shopify/polaris/pull/7691/files) release, I recommend we hold off on shipping this until v5 to avoid confusion with the migration, because running this script with v4 and v5 yield different results**

### WHY are these changes introduced?

Fixes #7504

When we turn the stylelint-polaris linter on in codebases, we need a way to ignore current warnings and errors since most if not all of these failures would require some sort of refactor (there isn't a one-to-one mapping of fixes).

This will ensure that refactors do not block us from turning on the linter in order to "stop the bleeding" and prevent further regressions.

### WHAT is this pull request doing?

Creates a new styles migration called `styles-insert-stylelint-disable` that goes through the given sass files and inserts a disable comment like:

```diff
+ // stylelint-disable-next-line -- generated by polaris-migrator DO NOT COPY
padding: 1rem;
```

There is also logic to deal with edge cases like if there is already an existing `stylelint-dsable-next-line` comment, it will merge the descriptions. See more about the edge cases in the tests, or in the `polaris-react` draft pr below (I left comments with some examples)

### How to 🎩
* Test on `polaris-react` https://github.com/Shopify/polaris/pull/7773 (3140 failures -> 4 failures)
* Test on `shopify/web` https://github.com/Shopify/web/pull/78162 (30383 failures -> 9 failures)

1. if testing in polaris-react, turn off the `overrides` in `.stylelintrc.js`
2. run `npx stylelint 'polaris-react/**/*.scss'` and note all the failures
3. run `yarn build && node ./bin/migrator-cli.js styles-insert-stylelint-disable 'polaris-react/**/*.scss'`
4. run `npx stylelint 'polaris-react/**/*.scss'` and note the reduced failures

If using a snapshot, run the migrator with `npx @shopify/polaris-migrator styles-insert-stylelint-disable <path>`. The migrator just uses a simple `@shopify/polaris-stylelint` config, so if double checking with `npx stylelint` make sure you are running the same version with the same config

### CODE REVIEW

File: polaris-migrator/src/migrations/styles-insert-stylelint-disable/styles-insert-stylelint-disable.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        if (node.prev()?.text?.includes(`-- polaris:`)) {
+          return;
+        }
+
+        // If the node is disable comment missing a polaris context message,
+        // append a generated context message
+        if (failureMsg.includes('-- polaris:')) {
+          node.text = `${node.text} ${polarisContextMsg}`;
+          return;
+        }
```
Comment (qt314) [inline polaris-migrator/src/migrations/styles-insert-stylelint-disable/styles-insert-stylelint-disable.ts:27]: this will never be true until [stylelint-polaris v5](https://github.com/Shopify/polaris/pull/7691/files) is shipped so I couldn't write a test for it. 

I could also omit the `if (failureMsg.includes('-- polaris:')` and assume all nodes that are comments that are failing need to be appended with the context string, then I could put the test back in

File: polaris-migrator/src/migrations/styles-insert-stylelint-disable/styles-insert-stylelint-disable.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+      const stylelintResult = (result as any)
+        ?.stylelint as StylelintPostcssResult;
+
+      if (stylelintResult?.config?.reportDescriptionlessDisables) {
+        root.walkComments((comment) => {
+          if (!comment.text.includes('--')) {
+            comment.text = [comment.text, polarisContextMsg].join(' ');
+          }
+        });
+      }
```
Comment (qt314) [inline polaris-migrator/src/migrations/styles-insert-stylelint-disable/styles-insert-stylelint-disable.ts:24]: I couldn't write a test for this case yet because it will only be enabled once [stylelint-polaris v5](https://github.com/Shopify/polaris/pull/7691/files) ships. I tested manually though on a v5 branch

File: polaris-migrator/src/migrations/styles-insert-stylelint-disable/styles-insert-stylelint-disable.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    },
+  };
+};
+
+export default async function insertStylelintDisable(file: FileInfo) {
+  return postcss([
+    stylelint({
+      config: {
+        extends: ['@shopify/stylelint-polaris'],
+        reportDescriptionlessDisables: true,
```
Comment (sophschneider) [inline polaris-migrator/src/migrations/styles-insert-stylelint-disable/styles-insert-stylelint-disable.ts:54]: I'm forcing this for now so that I could write a test, but once [stylelint-polaris v5](https://github.com/Shopify/polaris/pull/7691/files) ships we can remove this since it will be part of the `@shopify/stylelint-polaris` config

File: polaris-migrator/src/migrations/styles-insert-stylelint-disable/styles-insert-stylelint-disable.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+        deleteExtraNewlinesBeforeNode(node);
+      });
+
+      fillDescriptionlessDisables(result, root);
+    },
+  };
+};
+
+export default async function insertStylelintDisable(file: FileInfo) {
```
Comment (sam-b-rose) [inline polaris-migrator/src/migrations/styles-insert-stylelint-disable/styles-insert-stylelint-disable.ts]:
```suggestion
export default async function stylesInsertStylelintDisable(file: FileInfo) {
```

**Review by sam-b-rose — APPROVED** [review]
This is awesome! Great job testing it out in web and polaris-react 💯

File: polaris-migrator/src/migrations/styles-insert-stylelint-disable/styles-insert-stylelint-disable.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+/**
+ * Reduces the number of newline characters in a node's before
+ * raws to just one. This is helpful to ensure that there is
+ * only one newline between a disable-next-line comment and
+ * the warning node.
+ */
+function deleteExtraNewlinesBeforeNode(node: Node) {
+  node.raws.before = `\n${node.raws.before.replace(
```
Comment (aaronccasanova) [inline polaris-migrator/src/migrations/styles-insert-stylelint-disable/styles-insert-stylelint-disable.ts]: Can we insert a cross platform new line here? https://nodejs.org/api/os.html#os_os_eol

File: polaris-migrator/src/migrations/styles-insert-stylelint-disable/styles-insert-stylelint-disable.ts [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        const commentText = `${
+          isMultiline ? 'stylelint-disable' : 'stylelint-disable-next-line'
+        } ${polarisContextMsg}`;
+
+        const comment = createDisableComment(commentText, node.prev());
+
+        node.before(comment);
+
+        if (isMultiline) {
+          node.type === 'atrule'
```
Comment (aaronccasanova) [inline polaris-migrator/src/migrations/styles-insert-stylelint-disable/styles-insert-stylelint-disable.ts]: Should we check selectors as well? e.g. https://postcss.org/api/#rule

**Review by aaronccasanova — APPROVED** [review]
